### ISSUE TITLE
cli testing

### ISSUE THREAD

**Issue by kwstannard** [JLWolfe1990/andros#1]
Hey Josh. Thought I would go ahead and share how I did the cli tests since I went through exactly what you went through with the `gets` before in a prior project.

`bin/play` sets up a game and then runs a tight repl, passing user input to a Game object. The test runs `bin/play` in a separate process and passes io via the pipe named x. Game object public methods always return an array of strings, kind of like Rack.

spec/play_spec.rb
``` ruby
RSpec.describe 'bin/play' do
  context 'when a user closes the input' do
    it 'exits with successfully' do
      x=IO.popen('bin/play', 'w+')
      pid=x.pid
      x.close

      expect($?).to be_success
      expect($?.pid).to eq(pid)
    end
  end

  context 'when a user provides input' do
    it 'plays a game' do
      inputs = %w(join play_round)
      x=IO.popen('bin/play spec/test_decks.yaml', 'w+')
      x.puts inputs
      expected = (Game.new.new + Game.new.join + Game.new.play_round).map{|i| i + "\n"}

      expect(x.readlines).to eq(expected)

      x.close
    end
  end
end
```

bin/play
``` ruby
#! /usr/bin/env ruby

# ...
game = Game.new(deck)
puts game.new
STDIN.each do |line|
  next unless game.respond_to?(line.chomp)

  puts game.public_send(line.chomp)

  break if game.over?
end
```

Regards,
Kelly

**Comment by JLWolfe1990** [comment]
OIC. That pid structure makes sense. Something else that was interesting about my solution was that I would've had to create a way to stub out the draw method as well. Maybe a fixture similar to what you did would've worked for that.

Thank you for sharing :)

**Comment by kwstannard** [comment]
Yeah, the fixture I think is better than stubs because it lines up the ability to save and load state pretty well.

I think I am actually going to save this CLI testing pattern because I did spend too much time trying to figure out how to read the read pipe without closing the write pipe before deciding it wasn’t worth it.
